### PR TITLE
[SPARK-41333][SPARK-41737] Implement `GroupedData.{min, max, avg, sum}`

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Implement `GroupedData.{min, max, avg, sum}`

### Why are the changes needed?
TLDR, `df.groupby().min` != `df.groupby().agg(min)`

1, `GroupedData.min` only accepts string typed column names;
2, if no columns are provided, `GroupedData.min` will check the schema and choose the columns with matched datatypes;
3, if a column with unacceptable datatype is given, than `GroupedData.min` will throw exception like

`"department" is not a numeric column. Aggregation function can only be applied on a numeric column.`

### Does this PR introduce _any_ user-facing change?
yes, make `GroupedData.{min, max, avg, sum}` consistent with PySpark

### How was this patch tested?
added UT

### CODE REVIEW

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon @grundprinzip

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 1]
```diff
@@ -1061,6 +1068,81 @@ class SparkConnectPlanner(session: SparkSession) {
     }
   }
 
+  private def transformNumericAggregate(agg: proto.Aggregate): LogicalPlan = {
```
Comment (grundprinzip) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:1071]: Is that duplication really necessary? What exactly is the difference on the server side for the numeric aggregation? Do we have this special handling in Catalyst for this special case?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 2]
```diff
@@ -1061,6 +1068,81 @@ class SparkConnectPlanner(session: SparkSession) {
...
+    if (!agg.hasInput) {
+      throw InvalidPlanInput("Aggregate needs a plan input")
+    }
+    if (agg.getAggregateExpressionsCount == 0) {
```
Comment (grundprinzip) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:1075]: wouldn't this be caught by the analyzer?

File: python/pyspark/sql/connect/group.py [hunk 3]
```diff
@@ -97,36 +97,46 @@ def agg(self, *exprs: Union[Column, Dict[str, str]]) -> "DataFrame":
...
-            scalar_function(fun, col(param)) if isinstance(param, str) else param,
-        ]
+    def _numeric_agg(self, function: str, cols: Sequence[str]) -> "DataFrame":
+        from pyspark.sql.connect.dataframe import DataFrame
+
+        assert isinstance(function, str) and function in ["min", "max", "avg", "sum"]
+
+        assert isinstance(cols, list) and all(isinstance(c, str) for c in cols)
+
+        return DataFrame.withPlan(
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/group.py:110]: what happens when you only do this change but leave the `is_mumeric` out?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 1 above)
Comment (zhengruifeng) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:1071]: I believe it's necessary to have a separate method for it, actually at first I attempted to implement it in existing `transformAggregate` then I found that the two paths are too different:
1, for the regular agg, we build the logical plan;
2, for the numeric agg, we have to use the dataframe API to avoid doing the analysis by myself.

Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:1071]: Can't you just let the analysis fail? It will fail anyways if you try to aggregate on non numeric columns.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 2 above)
Comment (zhengruifeng) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:1075]: this protocol `functionName: String, aggCols: Seq[String]` is just defined in this PR, we have to process it by ourself

File: python/pyspark/sql/connect/group.py (hunk 3 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/group.py:110]: the agg expressions are just literal strings, it will return something like:

```
In [6]: df.groupBy("age").agg(F.lit("max"), F.lit("col1")).show()
+---+---+----+
|age|max|col1|
+---+---+----+
|  2|max|col1|
|  5|max|col1|
+---+---+----+

```

Comment (grundprinzip) [reply python/pyspark/sql/connect/group.py:110]: Yeah, but why not convert the function to a scalar function like we do for the above agg method?

Comment (grundprinzip) [reply python/pyspark/sql/connect/group.py:110]: You're already validating the input anyways.

Comment (grundprinzip) [reply python/pyspark/sql/connect/group.py:110]: Wait, sorry my bad.

Comment (grundprinzip) [reply python/pyspark/sql/connect/group.py:110]: I see now why you do it that way, but I'm not convinced it's the best way. 

The only delta I see that can't be handled reasonably is the case of empty agg expressions where we know the function. And here I'm wondering if we should just get the schema instead.

Comment (zhengruifeng) [reply python/pyspark/sql/connect/group.py:110]: sound reasonable, we can analyze the schema in the client, then do not need to change the sever side. Let me take another look

**Review by grundprinzip — APPROVED** [review]

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by zhengruifeng** [comment]
thank you for the reviews @grundprinzip @HyukjinKwon
